### PR TITLE
Add support for vpcConnector and vpcConnectorEgressSettings customization for functions

### PR DESCRIPTION
**Pull request by pcboy**
### Description

Supports firebase-functions SDK feature in https://github.com/firebase/firebase-functions/pull/752

```js
functions
      .runWith({
        vpcConnector: 'test-connector',
        vpcConnectorEgressSettings: 'PRIVATE_RANGES_ONLY'
      })
      .auth.user()
      .onCreate((user) => user);

```

It's an important feature for people who want to do a lot more with their firebase cloud functions. For instance connecting their cloud functions to a cloudSQL database.  
Check issue [#552 in firebase-functions](https://github.com/firebase/firebase-functions/issues/552) to see that I'm not the only one in need for that feature. Sadly that ticket got closed.

### Scenarios Tested

I tested creating a new cloud function with the vpcConnector and the vpcConnectorEgressSettings, it works. I also tried updating a function and works fine too. I also tested without mentioning the vpcConnectorEgressSettings setting and it worked too (will be set as ALL_TRAFFIC by default.)

### Sample Commands

No actual change in CLI options. Simply configure the vpcConnector inside your firebase.json file.

### CODE REVIEW

**Comment by pcboy** [comment]
@googlebot I signed it!

**Comment by mbleigh** [comment]
@pcboy thanks for the PR! I do think this is a worthwhile feature to investigate adding, but I don't think adding it to `firebase.json` config is where I'd want to expose it. I think it's more appropriate for the [`runWith()`](https://firebase.google.com/docs/functions/manage-functions#set_timeout_and_memory_allocation) options in the firebase-functions SDK.

All SDK/API changes for Firebase have to go through an internal review process before they can be added, and I'm willing to represent this feature internally. I can't promise that we'll be able to dedicate much engineering time to it soon, so if you're amenable to creating a corresponding PR in the `firebase-functions` repo to add this to `runWith` and then update this PR to use that instead of `firebase.json` config, I think we can work toward getting this done.

Cheers!

**Comment by pcboy** [comment]
@mbleigh Agreed. It seems better to put it in runWith().  
I added the PR on the firebase-functions repository.  

https://github.com/firebase/firebase-functions/pull/752

File: src/gcp/cloudfunctions.js [hunk 1]
```diff
@@ -65,13 +65,19 @@ function _createFunction(options) {
...
     sourceUploadUrl: options.sourceUploadUrl,
     name: func,
     entryPoint: options.entryPoint,
     labels: options.labels,
     runtime: options.runtime,
   };
+
+  if (options.vpcConnector) {
+    data.vpcConnector = options.vpcConnector;
+    data.vpcConnectorEgressSettings = options.vpcConnectorEgressSettings || "ALL_TRAFFIC";
```
Comment (mbleigh) [inline src/gcp/cloudfunctions.js:79]: As with the change in `firebase-functions`, we should allow the server-side default here.

```suggestion
    if (options.vpcConnectorEgressSettings) {
      data.vpcConnectorEgressSettings = options.vpcConnectorEgressSettings;
    }
```

Comment (pcboy) [reply src/gcp/cloudfunctions.js]: Good catch, forgot this one. It's fixed.

File: src/gcp/cloudfunctions.js [hunk 2]
```diff
@@ -65,13 +65,21 @@ function _createFunction(options) {
...
     sourceUploadUrl: options.sourceUploadUrl,
     name: func,
     entryPoint: options.entryPoint,
     labels: options.labels,
     runtime: options.runtime,
   };
+
+  if (options.vpcConnector) {
+    data.vpcConnector = options.vpcConnector;
+  }
```
Comment (mbleigh) [inline src/gcp/cloudfunctions.js:79]:
```suggestion
  if (options.vpcConnector) {
    data.vpcConnector = options.vpcConnector;
    // use implied project/location if only given connector id
    if (!data.vpcConnector.includes("/")) {
      data.vpcConnector = `${location}/connectors/${data.vpcConnector}`;
    }
  }
```

One last update: as part of the internal API review, we want to make the `vpcConnector` something that can be defined in a project-agnostic way, so if it doesn't contain a `/` (i.e. is not of the form `projects/*/locations/*/connectors/*`) we will automatically fill in the project and location based on the current function.

Could you apply the same logic to updates as well?

Comment (pcboy) [reply src/gcp/cloudfunctions.js]: OK. Updated. :+1:

**Comment by laurenzlong** [comment]
Thanks so much for this contribution! We'll try to get it out tomorrow. Also I edited the description to no longer mention firebase.json so it's not confusing. (I got confused at first when reviewing)

**Comment by leandroIA** [comment]
Hello..
I've been following here, I'm in doubt if this is the right channel to ask.
I saw that they made the updates, but I installed in my code, but when I do the deployment and as if nothing was happening.
It continues to create the functions without the VPC set.

**Comment by mbleigh** [comment]
You will need to make sure you are on version `v3.11.0` or later of `firebase-functions` and version `v8.9.0` or later of `firebase-tools`. That's the most likely reason you're not seeing anything happen.

**Comment by leandroIA** [comment]
thanks for your reply.

I am using exactly these versions.


but when I do the deployment I verify that it is not configured in my cloud
function the vpcconector

Em seg, 24 de ago de 2020 19:01, Michael Bleigh <notifications@github.com>
escreveu:

**Comment by minism** [comment]
We're also seeing the same issue that @leandroIA is mentioning.  I have vpcConnector configured which results in no change to the deployed cloud function.

**Comment by stijnlw** [comment]
@mbleigh  I have the same problem as @leandroIA - the extra code (runWith) doesn't seem to do anything at all. The connector is not being attached to th function.

`exports.testMysql = functions.runWith({
    vpcConnector: 'vpc-conn-euw1',
    vpcConnectorEgressSettings: 'PRIVATE_RANGES_ONLY'
}).region('europe-west1')
    .https.onCall(async (data, context) => {
        return {'status':'blablabla'};
    })
`
result:

afbeelding

and this is with a brand new 'firebase-tools' install, like so:

`    # Install dependencies
    - apt-get update
    - apt-get -y install apt-transport-https ca-certificates gnupg curl sudo gcc g++ make rsync git
    # Add nodejs for firebase deploy
    - curl -sL https://deb.nodesource.com/setup_12.x | sudo -E bash -
    - apt-get update && apt-get install -y nodejs
    # Install firebase tools
    - npm install -g firebase-tools`

firebase -V reports 8.19.0

I don't know how to check version of firebase-functions

**Comment by mbleigh** [comment]
To check `firebase-functions` version look in the `package-lock.json` in your `functions` folder.

**Comment by stijnlw** [comment]
Thanks a lot for the quick reply, I was on 3.7.0 - now that I'm on 3.13.0 it works as expected. Too bad the firebase deploy doesn't throw an error when you're trying to do things it doesn't support, but it certainly was an easy 'fix'

**Pull request merged**